include/envoy/ext_authz/ext_authz.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/common/optional.h"
+#include "envoy/common/pure.h"
+#include "envoy/http/filter.h"
+#include "envoy/http/header_map.h"
+#include "envoy/network/filter.h"
+#include "envoy/tracing/http_tracer.h"
+
+#include "api/auth/external_auth.pb.h"
+
+namespace Envoy {
+namespace ExtAuthz {
+
+/**
+ * Possible async results for a check call.
+ */
+enum class CheckStatus {
+  // The request is authorized.
+  OK,
+  // The authz service could not be queried.
+  Error,
+  // The request is denied.
+  Denied
+};
+
+/**
+ * Async callbacks used during check() calls.
+ */
+class RequestCallbacks {
+public:
+  virtual ~RequestCallbacks() {}
+
+  /**
+   * Called when a check request is complete. The resulting status is supplied.
+   */
+  virtual void complete(CheckStatus status) PURE;
+};
+
+class Client {
+public:
+  // Destructor
+  virtual ~Client() {}
+
+  /**
+   * Cancel an inflight Check request.
+   */
+  virtual void cancel() PURE;
+
+  // A check call.
+  virtual void check(RequestCallbacks& callback, const envoy::api::v2::auth::CheckRequest& request,
+                     Tracing::Span& parent_span) PURE;
+};
+
+typedef std::unique_ptr<Client> ClientPtr;
+
+/**
+ * An interface for creating a  external authorization client.
+ */
+class ClientFactory {
+public:
+  virtual ~ClientFactory() {}
+
+  /**
+   * Return a new authz client.
+   */
+  virtual ClientPtr create(const Optional<std::chrono::milliseconds>& timeout) PURE;
+};
+
+typedef std::unique_ptr<ClientFactory> ClientFactoryPtr;
+
+/**
+ * An interface for creating ext_authz.proto (authorization) request.
+ * CheckRequestGenerator is used to extract attributes from the TCP/HTTP request
+ * and fill out the details in the authorization protobuf that is sent to authorization
+ * service.
+ */
+class CheckRequestGenerator {
```

Comment:
This is really a helper utility class. Generally we don't put these in `include/envoy`. A litmus test is "will I need to mock this class?" and "will this class have potentially more than one implementation?". If no to both, put it in the `source/common` subtree.

Reply:
Initially I had it in the `souce/common` subtree. When I started writing the UT's for the filters I had to [mock](https://github.com/colabsaumoh/envoy/blob/141a203e47df1ba097e056b9395d8d47a0864f3b/test/mocks/ext_authz/mocks.h#L25) it. That's when I moved it here.
The usage in http filter is [here](
https://github.com/colabsaumoh/envoy/blob/fe93fd61b6ebb73247b089f47c7ea48d75e6af73/test/common/http/filter/ext_authz_test.cc#L117)

I think that the mock is needed. @htuch Should I just keep it here or move to `source/common` and have the mock reference it from there? please confirm. thx.

Reply:
Do you really need to mock the proto creation? These are basically straight line pure functions. Personally, I would mock the input parameters, since it's just as easy to do and results in easier to understand code.

Reply:
i think i know what you mean here. Let me try to redo the ut's without using the mock.

Reply:
@htuch last commit removes the mock for proto creation.